Support multiple cores

therealfrauholle opened the pull request:
This PR makes the tool aware of multiple cores. I also started to refactor the `mcd` library, which is further continued in https://github.com/therealfrauholle/tricore-probe/tree/support-multiple-chips.

Commit:
Wrap query function for core in safe rust function

Commit:
Query information for all cores

Commit:
Keep track of all cores in the structure

Commit:
Halt execution also if another core stops

Commit:
Check for correct flag

Commit:
Add better error reporting

Commit:
Implement better error handling

Commit:
Allow querying core state information gracefully

Commit:
Add log statement

Commit:
Also reset secondary cores

Commit:
Implement best effort to collect all data from RTT block

Commit:
Move log statement to debug level

Commit:
Improve core index report

Commit:
Also wait on main core when debug is hit

arctic-alpaca commented on rust-mcd/src/core.rs:
```diff
@@ -61,15 +61,30 @@ impl<'a> Core<'a> {
-            return Err(expect_error(Some(self))).with_context(|| "Could not query device state");
+            let error = expect_error(Some(self));
+            
+            if error.event_error_code().intersects(tolerate_events) {
```
Here `Ok(None)` is returned if the actual error matches the tolerated events/errors. Is it intended to allow any other error in that case too?
E.g. error bitflags: `0111`
tolerated bitflags: `0100`
This would return `Ok(None)` even though there were non-tolerated errors.

Follow up question:
Is EventError guaranteed to have only one of the bitflags set?

therealfrauholle replied:
Quoting from the documentation of `mcd_error_event_et`

See https://github.com/veecle/tricore-probe/blob/6a55de0ba4d6a4e468e33741484b2b63a92f6de2/rust-mcd/mcd_demo_basic_120412/src/mcd_api.h#L302

For the sake of this function, I implemented that type as bitflags, but I now see that it is overall better to use a normal rust enum here, what do you think?

arctic-alpaca replied:
I agree with the enum approach :+1:

arctic-alpaca commented on rust-mcd/src/raw/query_core.rs:
```diff
@@ -0,0 +1,61 @@
+    ) -> Result<Vec<mcd_core_con_info_st>> {
+        assert!(
+            core_query_count > 0,
+            "Can only query positive number of cores"
```
```suggestion
            "Can only query non-zero number of cores"
```

arctic-alpaca commented on rust-mcd/src/error.rs:
```diff
@@ -31,14 +31,28 @@ pub struct Error {
 
 impl Display for Error {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         let info = unsafe { CStr::from_ptr(&self.inner.error_str[0] as *const i8) }
```
This could be changed to avoid the invariants `CStr::from_ptr()` requires:
```rust
// SAFETY:
// Arrays of u8 share the same layout as arrays of i8 and i8 can be transmuted to u8 safely.
let error_str: &[u8; 256] = unsafe { std::mem::transmute(&self.inner.error_str) };
let info = CStr::from_bytes_until_nul(error_str)
    .unwrap()
    .to_str()
    .unwrap();
```

Depending on performance needs, this could also be achieved without unsafe via `map()`.

therealfrauholle replied:
Good point!

arctic-alpaca commented on rust-mcd/src/raw/query_core.rs:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    /// See [Self::mcd_qry_cores_f], with core_count set to 0
```
I think the comment does not match the function as it returns the core count and not `mcd_core_con_info_st`.

therealfrauholle replied:
`mcd_qry_cores_f` has two call patterns:

1. Obtain the number of available cores (set `num_devices` to 0)
2. Obtain core information (set `num_devices` > 0)

I clarified the documentation.

arctic-alpaca commented on tricore-windows/src/defmt.rs:
```diff
@@ -4,27 +4,38 @@ use byteorder::ReadBytesExt;
 /// The backtrace returned is obtained by traversing the CSA link list.
 pub fn decode_rtt<W: Write>(
     core: &mut Core<'_>,
+    secondary_cores: &mut [Core<'_>],
```
What are `secondary_cores` and what's their function?

therealfrauholle replied:
They are observed for their status, in case they hit a breakpoint (e.g. a panic). I clarified the function documentation.

arctic-alpaca commented on tricore-windows/src/defmt.rs:
```diff
@@ -85,55 +90,94 @@ pub fn decode_rtt<W: Write>(
+        /// Check if the core is still running, if it is not we assume a
+        /// breakpoint was hit
+        ///
+        /// This function is a bit of a hacky to work around lifetime issues
```
```suggestion
        /// This function is a bit of a hack to work around lifetime issues
```

arctic-alpaca commented on tricore-windows/src/defmt.rs:
```diff
@@ -85,55 +90,94 @@ pub fn decode_rtt<W: Write>(
+        for (secondary_index, core) in secondary_cores.iter_mut().enumerate() {
+            if let Some(exit_reason) = should_exit_fore_core(core, true) {
+                if exit_reason.is_ok() {
+                    // FIXME: The core index we give here might be misleading, we can probably obtain
```
FIXME in PR.

therealfrauholle replied:
Lets leave this in for now, I expect us to change this code a lot in the close future.

arctic-alpaca commented on tricore-windows/src/lib.rs:
```diff
@@ -52,8 +61,9 @@ impl ChipInterface {
-        let core = system.get_first_core()?;
+        let core = system.get_core(0)?;
         let system_reset = ResetClass::construct_reset_class(&core, 0);
+        // Do we also need to reset the other cores?
```
Should we find this out before or after merge?

therealfrauholle replied:
For now no.

I am not familiar with reset classes of the TC* chips, e.g. there are different reset "types". This knowledge will come with further work on this tool and with the library itself, I already spend a bit of time wrapping my head around what cores we need to reset.

The same applies for what "downloading triggers" means for an actual device when creating a breakpoint...

arctic-alpaca commented on tricore-windows/src/defmt.rs:
```diff
@@ -4,27 +4,38 @@ use byteorder::ReadBytesExt;
     // Can we assume that this reset class exists and it fits the usecase here?
     let system_reset = ResetClass::construct_reset_class(core, 0);
+
+    // Do we also need to reset the other cores?
```
Do we find this out before or after merge?

therealfrauholle replied:
See https://github.com/veecle/tricore-probe/pull/3#discussion_r1479730018

Commit:
Sanity check for null-byte in string

Commit:
Clarify documentation

Commit:
Clarify meaning of secondary cores

Commit:
Fix typo

Commit:
Improve crash message

Commit:
Make EventError normal rust enumeration

Event flags are bitwise exclusive (according to the documentation), so a rust enum makes more sense here.

Commit:
Re-add documentation for error

arctic-alpaca reviewed (approved):
Thanks for addressing everything. Looks good to me :+1:
